```
Error running mint automation
+ mkdir -p 15923-7425e58/gopath/src/github.com/minio
+ git -C 15923-7425e58/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 15923-7425e58/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 15923-7425e58/gopath/src/github.com/minio/minio fetch --quiet minio pull/15923/head:pr15923
+ git -C 15923-7425e58/gopath/src/github.com/minio/minio checkout --quiet pr15923
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/15923-7425e58/gopath
+ make -C 15923-7425e58/gopath/src/github.com/minio/minio --quiet
go: downloading github.com/minio/madmin-go v1.6.5
go: downloading github.com/Shopify/sarama v1.36.0
go: downloading github.com/xdg/scram v1.0.5
go: downloading github.com/fatih/color v1.13.0
go: downloading github.com/minio/cli v1.24.0
go: downloading github.com/minio/pkg v1.5.2
go: downloading github.com/minio/highwayhash v1.0.2
go: downloading github.com/minio/sio v0.3.0
go: downloading github.com/json-iterator/go v1.1.12
go: downloading github.com/klauspost/compress v1.15.11
go: downloading github.com/minio/sha256-simd v1.0.0
go: downloading github.com/rs/dnscache v0.0.0-20211102005908-e0241e321417
go: downloading github.com/secure-io/sio-go v0.3.1
go: downloading github.com/golang-jwt/jwt/v4 v4.4.2
go: downloading github.com/dustin/go-humanize v1.0.0
go: downloading golang.org/x/sys v0.0.0-20220915200043-7b5979e65e41
go: downloading github.com/buger/jsonparser v1.1.1
go: downloading golang.org/x/crypto v0.0.0-20221005025214-4161e89ecf1b
go: downloading github.com/minio/kes v0.21.0
go: downloading cloud.google.com/go/storage v1.26.0
go: downloading github.com/lithammer/shortuuid/v4 v4.0.0
go: downloading github.com/Azure/azure-storage-blob-go v0.15.0
go: downloading github.com/gorilla/mux v1.8.0
go: downloading github.com/cespare/xxhash/v2 v2.1.2
go: downloading github.com/klauspost/pgzip v1.2.5
go: downloading github.com/bcicen/jstream v1.0.1
go: downloading github.com/bits-and-blooms/bloom/v3 v3.3.1
go: downloading github.com/google/uuid v1.3.0
go: downloading github.com/cheggaaa/pb v1.0.29
go: downloading github.com/inconshreveable/mousetrap v1.0.1
go: downloading github.com/minio/console v0.21.0
go: downloading github.com/dchest/siphash v1.2.3
go: downloading github.com/klauspost/reedsolomon v1.11.0
go: downloading github.com/coreos/go-oidc v2.2.1+incompatible
go: downloading github.com/klauspost/readahead v1.4.0
go: downloading github.com/djherbis/atime v1.1.0
go: downloading github.com/hashicorp/golang-lru v0.5.4
go: downloading github.com/go-openapi/loads v0.21.2
go: downloading github.com/miekg/dns v1.1.50
go: downloading github.com/cosnicolaou/pbzip2 v1.0.1
go: downloading github.com/minio/xxml v0.0.3
go: downloading github.com/felixge/fgprof v0.9.3
go: downloading github.com/minio/zipindex v0.3.0
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading github.com/minio/selfupdate v0.5.0
go: downloading github.com/pierrec/lz4 v2.6.1+incompatible
go: downloading github.com/philhofer/fwd v1.1.2-0.20210722190033-5c56ac6d0bb9
go: downloading github.com/prometheus/client_golang v1.13.0
go: downloading github.com/prometheus/client_model v0.2.0
go: downloading github.com/prometheus/common v0.37.0
go: downloading github.com/rs/cors v1.8.2
go: downloading github.com/minio/dperf v0.4.2
go: downloading github.com/prometheus/procfs v0.8.0
go: downloading github.com/valyala/bytebufferpool v1.0.0
go: downloading github.com/tinylib/msgp v1.1.7-0.20220719154719-f3635b96e483
go: downloading github.com/shirou/gopsutil/v3 v3.22.8
go: downloading github.com/yargevad/filepathx v1.0.0
go: downloading github.com/zeebo/xxh3 v1.0.2
go: downloading go.etcd.io/etcd/api/v3 v3.5.5
go: downloading go.etcd.io/etcd/client/v3 v3.5.5
go: downloading go.uber.org/atomic v1.10.0
go: downloading golang.org/x/oauth2 v0.0.0-20220909003341-f21342109be1
go: downloading google.golang.org/api v0.96.0
go: downloading gopkg.in/yaml.v2 v2.4.0
go: downloading github.com/beevik/ntp v0.3.0
go: downloading go.uber.org/zap v1.23.0
go: downloading golang.org/x/time v0.0.0-20220722155302-e5dcc9cfc0b9
go: downloading github.com/coredns/coredns v1.9.4
go: downloading github.com/minio/simdjson-go v0.4.2
go: downloading github.com/go-ldap/ldap/v3 v3.4.4
go: downloading github.com/eclipse/paho.mqtt.golang v1.4.1
go: downloading github.com/elastic/go-elasticsearch/v7 v7.17.1
go: downloading github.com/go-sql-driver/mysql v1.6.0
go: downloading github.com/gomodule/redigo v1.8.9
go: downloading github.com/lib/pq v1.10.7
go: downloading github.com/nats-io/nats.go v1.17.0
go: downloading github.com/nats-io/stan.go v0.10.3
go: downloading github.com/nsqio/go-nsq v1.1.0
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/streadway/amqp v1.0.0
go: downloading github.com/ncw/directio v1.0.5
go: downloading github.com/mattn/go-runewidth v0.0.13
go: downloading github.com/golang/protobuf v1.5.2
go: downloading golang.org/x/sync v0.0.0-20220907140024-f12130a52804
go: downloading github.com/mattn/go-colorable v0.1.13
go: downloading github.com/mattn/go-isatty v0.0.16
go: downloading github.com/minio/csvparser v1.0.0
go: downloading github.com/fraugster/parquet-go v0.12.0
go: downloading github.com/bits-and-blooms/bitset v1.3.3
go: downloading github.com/alecthomas/participle v0.7.1
go: downloading github.com/pquerna/cachecontrol v0.1.0
go: downloading gopkg.in/square/go-jose.v2 v2.6.0
go: downloading github.com/xdg/stringprep v1.0.3
go: downloading github.com/klauspost/cpuid/v2 v2.1.1
go: downloading golang.org/x/net v0.0.0-20220909164309-bea034e7d591
go: downloading github.com/gogo/protobuf v1.3.2
go: downloading github.com/rivo/uniseg v0.4.2
go: downloading github.com/google/pprof v0.0.0-20220829040838-70bd9ae97f40
go: downloading github.com/Azure/go-ntlmssp v0.0.0-20220621081337-cb9428e4ac1e
go: downloading github.com/go-asn1-ber/asn1-ber v1.5.4
go: downloading go.etcd.io/etcd/client/pkg/v3 v3.5.5
go: downloading google.golang.org/grpc v1.49.0
go: downloading github.com/golang/snappy v0.0.4
go: downloading google.golang.org/genproto v0.0.0-20220916172020-2692e8806bfa
go: downloading github.com/coreos/go-semver v0.3.0
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading github.com/modern-go/reflect2 v1.0.2
go: downloading github.com/go-openapi/analysis v0.21.4
go: downloading github.com/go-openapi/spec v0.20.7
go: downloading github.com/go-openapi/swag v0.22.3
go: downloading golang.org/x/text v0.3.8
go: downloading github.com/gorilla/websocket v1.5.0
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.2-0.20181231171920-c182affec369
go: downloading github.com/beorn7/perks v1.0.1
go: downloading google.golang.org/protobuf v1.28.1
go: downloading github.com/coreos/go-systemd/v22 v22.4.0
go: downloading github.com/lestrrat-go/jwx v1.2.25
go: downloading github.com/montanaflynn/stats v0.6.6
go: downloading github.com/fatih/structs v1.1.0
go: downloading github.com/rjeczalik/notify v0.9.2
go: downloading github.com/Azure/azure-pipeline-go v0.2.3
go: downloading go.uber.org/multierr v1.8.0
go: downloading github.com/mailru/easyjson v0.7.7
go: downloading gopkg.in/yaml.v3 v3.0.1
go: downloading github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading github.com/eapache/go-resiliency v1.3.0
go: downloading github.com/eapache/queue v1.1.0
go: downloading cloud.google.com/go v0.104.0
go: downloading github.com/hashicorp/go-multierror v1.1.1
go: downloading github.com/jcmturner/gofork v1.7.6
go: downloading github.com/jcmturner/gokrb5/v8 v8.4.3
go: downloading github.com/pierrec/lz4/v4 v4.1.16
go: downloading github.com/rcrowley/go-metrics v0.0.0-20201227073835-cf1acfcdf475
go: downloading github.com/mattn/go-ieproxy v0.0.1
go: downloading github.com/apache/thrift v0.16.0
go: downloading github.com/nats-io/nkeys v0.3.0
go: downloading github.com/nats-io/nuid v1.0.1
go: downloading github.com/hashicorp/errwrap v1.1.0
go: downloading github.com/josharian/intern v1.0.0
go: downloading github.com/tklauser/go-sysconf v0.3.10
go: downloading github.com/jcmturner/dnsutils/v2 v2.0.0
go: downloading github.com/hashicorp/go-uuid v1.0.3
go: downloading github.com/jcmturner/aescts/v2 v2.0.0
go: downloading github.com/jcmturner/rpc/v2 v2.0.3
go: downloading github.com/go-openapi/jsonpointer v0.19.5
go: downloading github.com/go-openapi/strfmt v0.21.3
go: downloading github.com/lestrrat-go/backoff/v2 v2.0.8
go: downloading github.com/lestrrat-go/iter v1.0.2
go: downloading github.com/lestrrat-go/option v1.0.0
go: downloading github.com/lestrrat-go/blackmagic v1.0.1
go: downloading github.com/lestrrat-go/httpcc v1.0.1
go: downloading github.com/tklauser/numcpus v0.5.0
go: downloading github.com/go-openapi/jsonreference v0.20.0
go: downloading github.com/go-openapi/errors v0.20.3
go: downloading github.com/asaskevich/govalidator v0.0.0-20210307081110-f21760c49a8d
go: downloading github.com/oklog/ulid v1.3.1
go: downloading github.com/mitchellh/mapstructure v1.5.0
go: downloading go.mongodb.org/mongo-driver v1.10.2
go: downloading cloud.google.com/go/compute v1.10.0
go: downloading cloud.google.com/go/iam v0.4.0
go: downloading github.com/googleapis/gax-go/v2 v2.5.1
go: downloading golang.org/x/xerrors v0.0.0-20220907171357-04be3eba64a2
go: downloading go.opencensus.io v0.23.0
go: downloading github.com/golang/groupcache v0.0.0-20210331224755-41bb18bfe9da
go: downloading github.com/go-openapi/runtime v0.24.1
go: downloading github.com/go-openapi/validate v0.22.0
go: downloading github.com/jessevdk/go-flags v1.5.0
go: downloading github.com/minio/mc v0.0.0-20221007160339-ec8687d57e36
go: downloading github.com/rs/xid v1.4.0
go: downloading github.com/tidwall/gjson v1.14.3
go: downloading github.com/unrolled/secure v1.13.0
go: downloading github.com/tidwall/pretty v1.2.0
go: downloading github.com/tidwall/match v1.1.1
go: downloading github.com/docker/go-units v0.5.0
go: downloading github.com/charmbracelet/bubbles v0.14.0
go: downloading github.com/charmbracelet/bubbletea v0.22.1
go: downloading github.com/minio/colorjson v1.0.2
go: downloading github.com/charmbracelet/lipgloss v0.6.0
go: downloading gopkg.in/h2non/filetype.v1 v1.0.5
go: downloading github.com/google/shlex v0.0.0-20191202100458-e7afc7fbc510
go: downloading github.com/pkg/xattr v0.4.9
go: downloading github.com/muesli/reflow v0.3.0
go: downloading github.com/gdamore/tcell/v2 v2.5.3
go: downloading github.com/posener/complete v1.2.3
go: downloading golang.org/x/term v0.0.0-20220722155259-a9ba230a4035
go: downloading github.com/minio/filepath v1.0.0
go: downloading github.com/rivo/tview v0.0.0-20220916081518-2e69b7385a37
go: downloading github.com/olekukonko/tablewriter v0.0.5
go: downloading github.com/prometheus/prom2json v1.3.1
go: downloading github.com/navidys/tvxwidgets v0.1.1
go: downloading github.com/muesli/termenv v0.12.0
go: downloading github.com/lucasb-eyer/go-colorful v1.2.0
go: downloading github.com/gdamore/encoding v1.0.0
go: downloading github.com/muesli/cancelreader v0.2.2
go: downloading github.com/mattn/go-localereader v0.0.1
go: downloading github.com/muesli/ansi v0.0.0-20211031195517-c9f0611b6c70
go: downloading github.com/containerd/console v1.0.3
go: downloading github.com/googleapis/enterprise-certificate-proxy v0.1.0
go: downloading github.com/google/go-cmp v0.5.9
../../../../pkg/mod/github.com/minio/mc@v0.0.0-20221007160339-ec8687d57e36/cmd/alias-set.go:36:2: github.com/minio/minio-go/v7@v7.0.41-0.20221013203648-8257e7003b5e: replacement directory ../minio-go does not exist
../../../../pkg/mod/github.com/minio/console@v0.21.0/pkg/auth/idp/oauth2/provider.go:32:2: github.com/minio/minio-go/v7@v7.0.41-0.20221013203648-8257e7003b5e: replacement directory ../minio-go does not exist
../../../../pkg/mod/github.com/minio/mc@v0.0.0-20221007160339-ec8687d57e36/cmd/client-fs.go:43:2: github.com/minio/minio-go/v7@v7.0.41-0.20221013203648-8257e7003b5e: replacement directory ../minio-go does not exist
../../../../pkg/mod/github.com/minio/mc@v0.0.0-20221007160339-ec8687d57e36/cmd/ilm/options.go:28:2: github.com/minio/minio-go/v7@v7.0.41-0.20221013203648-8257e7003b5e: replacement directory ../minio-go does not exist
../../../../pkg/mod/github.com/minio/mc@v0.0.0-20221007160339-ec8687d57e36/cmd/client-fs.go:45:2: github.com/minio/minio-go/v7@v7.0.41-0.20221013203648-8257e7003b5e: replacement directory ../minio-go does not exist
../../../../pkg/mod/github.com/minio/mc@v0.0.0-20221007160339-ec8687d57e36/cmd/client-s3.go:49:2: github.com/minio/minio-go/v7@v7.0.41-0.20221013203648-8257e7003b5e: replacement directory ../minio-go does not exist
../../../../pkg/mod/github.com/minio/madmin-go@v1.6.5/cluster-commands.go:28:2: github.com/minio/minio-go/v7@v7.0.41-0.20221013203648-8257e7003b5e: replacement directory ../minio-go does not exist
../../../../pkg/mod/github.com/minio/madmin-go@v1.6.5/anonymous-api.go:36:2: github.com/minio/minio-go/v7@v7.0.41-0.20221013203648-8257e7003b5e: replacement directory ../minio-go does not exist
../../../../pkg/mod/github.com/minio/console@v0.21.0/pkg/auth/idp/oauth2/provider.go:33:2: github.com/minio/minio-go/v7@v7.0.41-0.20221013203648-8257e7003b5e: replacement directory ../minio-go does not exist
../../../../pkg/mod/github.com/minio/madmin-go@v1.6.5/api.go:42:2: github.com/minio/minio-go/v7@v7.0.41-0.20221013203648-8257e7003b5e: replacement directory ../minio-go does not exist
../../../../pkg/mod/github.com/minio/mc@v0.0.0-20221007160339-ec8687d57e36/cmd/client-s3.go:51:2: github.com/minio/minio-go/v7@v7.0.41-0.20221013203648-8257e7003b5e: replacement directory ../minio-go does not exist
../../../../pkg/mod/github.com/minio/madmin-go@v1.6.5/user-commands.go:27:2: github.com/minio/minio-go/v7@v7.0.41-0.20221013203648-8257e7003b5e: replacement directory ../minio-go does not exist
make: *** [build] Error 1

```

--- a/cmd/bucket-replication.go
+++ b/cmd/bucket-replication.go
@@ -580,8 +580,8 @@ func replicateDeleteToTarget(ctx context.Context, dobj DeletedObjectReplicationI
 		toi, err := tgt.StatObject(ctx, tgt.Bucket, dobj.ObjectName, miniogo.StatObjectOptions{
 			VersionID: versionID,
 			Internal: miniogo.AdvancedGetOptions{
-				ReplicationProxyRequest: "false",
-				ReplicationDeleteMarker: true,
+				ReplicationProxyRequest:           "false",
+				IsReplicationReadyForDeleteMarker: true,
 			},
 		})
 		if isErrMethodNotAllowed(ErrorRespToObjectError(err, dobj.Bucket, dobj.ObjectName)) {

--- a/cmd/object-api-interface.go
+++ b/cmd/object-api-interface.go
@@ -51,7 +51,9 @@ type ObjectOptions struct {
 	MTime                time.Time // Is only set in POST/PUT operations
 	Expires              time.Time // Is only used in POST/PUT operations
 
-	DeleteMarker      bool                // Is only set in DELETE operations for delete marker replication
+	DeleteMarker            bool // Is only set in DELETE operations for delete marker replication
+	CheckDMReplicationReady bool // Is delete marker ready to be replicated - set only during HEAD
+
 	UserDefined       map[string]string   // only set in case of POST/PUT operations
 	PartNumber        int                 // only useful in case of GetObject/HeadObject
 	CheckPrecondFn    CheckPreconditionFn // only set during GetObject/HeadObject/CopyObjectPart preconditional valuation

--- a/cmd/object-api-options.go
+++ b/cmd/object-api-options.go
@@ -159,6 +159,23 @@ func getOpts(ctx context.Context, r *http.Request, bucket, object string) (Objec
 			}
 		}
 	}
+	replReadyCheck := strings.TrimSpace(r.Header.Get(xhttp.MinIOCheckDMReplicationReady))
+	if replReadyCheck != "" {
+		switch replReadyCheck {
+		case "true":
+			opts.CheckDMReplicationReady = true
+		case "false":
+		default:
+			err = fmt.Errorf("Unable to parse %s, failed with %w", xhttp.MinIOCheckDMReplicationReady, fmt.Errorf("should be true or false"))
+			logger.LogIf(ctx, err)
+			return opts, InvalidArgument{
+				Bucket: bucket,
+				Object: object,
+				Err:    err,
+			}
+		}
+	}
+
 	opts.Versioned = globalBucketVersioningSys.PrefixEnabled(bucket, object)
 	opts.VersionSuspended = globalBucketVersioningSys.PrefixSuspended(bucket, object)
 	return opts, nil

--- a/cmd/object-handlers.go
+++ b/cmd/object-handlers.go
@@ -739,14 +739,15 @@ func (api objectAPIHandlers) headObjectHandler(ctx context.Context, objectAPI Ob
 				w.Header()[xhttp.AmzDeleteMarker] = []string{strconv.FormatBool(objInfo.DeleteMarker)}
 			}
 			QueueReplicationHeal(ctx, bucket, objInfo)
-		}
-		// do an additional verification whether object exists when opts.DeleteMarker is set by source
-		// cluster as part of delete marker replication
-		if opts.DeleteMarker && opts.ProxyHeaderSet {
-			opts.VersionID = ""
-			goi, gerr := getObjectInfo(ctx, bucket, object, opts)
-			if gerr == nil || goi.VersionID != "" { // object layer returned more info because object is deleted
-				w.Header().Set(xhttp.MinIOTargetReplicationReady, "true")
+			// do an additional verification whether object exists when object is deletemarker and request
+			// is from replication
+			if opts.CheckDMReplicationReady {
+				topts := opts
+				topts.VersionID = ""
+				goi, gerr := getObjectInfo(ctx, bucket, object, topts)
+				if gerr == nil || goi.VersionID != "" { // object layer returned more info because object is deleted
+					w.Header().Set(xhttp.MinIOTargetReplicationReady, "true")
+				}
 			}
 		}
 		writeErrorResponseHeadersOnly(w, toAPIError(ctx, err))

--- a/go.mod
+++ b/go.mod
@@ -49,7 +49,7 @@ require (
 	github.com/minio/highwayhash v1.0.2
 	github.com/minio/kes v0.21.0
 	github.com/minio/madmin-go v1.6.5
-	github.com/minio/minio-go/v7 v7.0.41-0.20221013203648-8257e7003b5e
+	github.com/minio/minio-go/v7 v7.0.43-0.20221021202758-c6319beb6b27
 	github.com/minio/pkg v1.5.2
 	github.com/minio/selfupdate v0.5.0
 	github.com/minio/sha256-simd v1.0.0

--- a/internal/http/headers.go
+++ b/internal/http/headers.go
@@ -197,6 +197,8 @@ const (
 	MinIOReplicationResetStatus = "X-Minio-Replication-Reset-Status"
 	// Header indicating target cluster can receive delete marker replication requests because object has been replicated
 	MinIOTargetReplicationReady = "X-Minio-Replication-Ready"
+	// Header asking if cluster can receive delete marker replication request now.
+	MinIOCheckDMReplicationReady = "X-Minio-Check-Replication-Ready"
 	// Header indiicates last tag update time on source
 	MinIOSourceTaggingTimestamp = "X-Minio-Source-Replication-Tagging-Timestamp"
 	// Header indiicates last rtention update time on source

